java/main/com/google/jglome/Glome.java

```diff
@@ -0,0 +1,277 @@
+    this.userPrivateKey = builder.userPrivateKey;
+    this.peerKey = builder.peerKey;
+    this.minPeerTagLength = builder.minPeerTagLength;
+    this.userMacKey = builder.userMacKey;
+    this.peerMacKey = builder.peerMacKey;
+  }
+
+  public byte[] peerKey() {
+    return peerKey;
+  }
+
+  public byte[] userPublicKey() {
+    return userPublicKey;
+  }
+
+  public byte[] userPrivateKey() {
+    return userPrivateKey;
+  }
+
+  /**
+   * Generates a user's tag corresponding to the given message {@code msg} and the counter {@code
+   * cnt}.
+   *
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the user to the peer.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [MIN_CNT_VALUE..MAX_CNT_VALUE]
+   * range.
+   */
+  public byte[] generateTag(byte[] msg, int cnt) throws CounterOutOfBoundsException {
+    return generateTag(msg, cnt, this.userMacKey);
+  }
+
+  /**
+   * Checks whether the peer's tag matches received message and some counter.
+   *
+   * @param peerTag tag from a peer.
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the user to the peer.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [MIN_CNT_VALUE..MAX_CNT_VALUE]
+   * range.
+   * @throws WrongTagException if the peer's tag has invalid length (less than {@code
+   * minPeerTagLength} or more than MAX_TAG_LENGTH) or it's not equal to the prefix of a correct
+   * tag.
+   */
+  public void checkTag(byte[] peerTag, byte[] msg, int cnt)
+      throws CounterOutOfBoundsException, WrongTagException {
+    if (peerTag.length < minPeerTagLength || peerTag.length > MAX_TAG_LENGTH) {
+      throw new WrongTagException(
+          String.format("The received tag has invalid length. Expected %d or more, got %d.",
+              minPeerTagLength, peerTag.length));
+    }
+
+    byte[] truncatedTag = Arrays.copyOf(generateTag(msg, cnt, this.peerMacKey), peerTag.length);
+    if (!Arrays.equals(peerTag, truncatedTag)) {
+      throw new WrongTagException("The received tag doesn't match the expected tag.");
+    }
+  }
+
+  /**
+   * Generates a tag corresponding to the given message {@code msg} and the counter {@code cnt}.
+   *
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the user to the peer.
+   * @param mac MAC key of a sender.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [MIN_CNT_VALUE..MAX_CNT_VALUE]
+   * range.
+   */
+  private byte[] generateTag(byte[] msg, int cnt, Mac mac)
+      throws CounterOutOfBoundsException {
+    if (cnt < MIN_CNT_VALUE || cnt > MAX_CNT_VALUE) {
+      throw new CounterOutOfBoundsException(
+          String
+              .format("Counter should be in [%d..%d] range. Got %d.", MIN_CNT_VALUE, MAX_CNT_VALUE,
+                  cnt));
+    }
+
+    byte[] finalMsg = new byte[msg.length + 1];
+    finalMsg[0] = (byte) cnt; // for [0..255] range do `& 0xFF`
```

Comment:
The range has already been asserted above, is the comment obsolete?

Reply:
It's just a reminder that `byte` in Java is having a special value interval [-128..127], while counter values are in [0..255] range.